SSA/Core/HVector.lean

```diff
@@ -222,5 +222,39 @@ instance [Lean.ToExpr α] [∀ a, Lean.ToExpr (A a)] [HVector.ToExprPi A]
 
 end ToExprPi
 
+/-!
+  ## Folding and mapping over HVectors, monadic and non-monadic.
+-/
+section mapHVec
+
+def foldlM {B : Type*} [Monad m] (f : ∀ (a : α), B → A a → m B) :
+    ∀ {l : List α}, (init : B) → (as : HVector A l) → m B
+  | [],   b, .nil       => return b
+  | t::_, b, .cons a as => do foldlM f (← f t b a) as
+
+/-! Simultaneous map on the type and value level of an HVector. -/
+def ubermap {A : α → Type} {B : β → Type}
+    {l : List α}
+    (F : α → β)
+    (f : {a : α} → (v : A a) → B (F a) )
+    (as : HVector A l) : (HVector B (F <$> l)) :=
+  match l, as with
+  | [], .nil => .nil
+  | _t :: _ts, .cons a as => HVector.cons (f a) (HVector.ubermap F f as)
+
+/-!
+Simultaneous map on the type and value level of an HVector while
+performing monadic effects for value translation.-/
+def ubermapM [Monad m] {α : Type 0} {A : α → Type} {B : β → Type}
```

Comment:
I don't think we have a `mapM'`, but if `ubermap` is indeed the same as `map'`, then we should name this `mapM'` to match.